crates/dbsp/src/circuit/dbsp_handle.rs

```diff
@@ -828,4 +961,412 @@ mod tests {
+    /// circuit with this storage directory.
+    #[test]
+    fn gc_on_startup() {
+        let _r = env_logger::try_init();
+
+        let (temp, cconf) = mkconfig();
+        let (mut dbsp, (input_handle, _, _)) = mkcircuit(&cconf).unwrap();
+
+        let mut batch: Vec<Tup2<i32, Tup2<i32, i64>>> = vec![Tup2(1, Tup2(2, 1))];
+        input_handle.append(&mut batch);
+        dbsp.commit().expect("commit shouldn't fail");
+        drop(dbsp);
+
+        let incomplete_batch_path = temp.path().join("incomplete_batch.mut");
+        let _ = File::create(&incomplete_batch_path).expect("can't create file");
+
+        let incomplete_checkpoint_dir = temp.path().join(Uuid::now_v7().to_string());
+        fs::create_dir(&incomplete_checkpoint_dir).expect("can't create checkpoint dir");
+
+        let complete_batch_unused = temp.path().join("complete_batch.feldera");
+        let _ = File::create(&complete_batch_unused).expect("can't create file");
+
+        // Initializing this circuit again will remove the
+        // unnecessary files in the checkpoint directory.
+        let (_dbsp, _) = mkcircuit(&cconf).unwrap();
+
+        assert!(!incomplete_checkpoint_dir.exists());
+        assert!(!incomplete_batch_path.exists());
+        assert!(!complete_batch_unused.exists());
+    }
+
+    /// Make sure we can take checkpoints of a simple spine and restore them.
+    #[test]
+    fn commit_restore() {
+        let _r = env_logger::try_init();
+        let batches: Vec<Vec<Tup2<i32, Tup2<i32, i64>>>> = vec![
+            vec![Tup2(1, Tup2(2, 1))],
+            vec![Tup2(3, Tup2(4, 1))],
+            vec![Tup2(5, Tup2(6, 1))],
+            vec![Tup2(7, Tup2(8, 1))],
+            vec![Tup2(9, Tup2(10, 1))],
+        ];
+
+        generic_checkpoint_restore(batches, mkcircuit);
+    }
+
+    /// Make sure we can take checkpoints of a spine with a trace bound and
+    /// restore them.
+    #[test]
+    fn commit_restore_bounds() {
+        let _r = env_logger::try_init();
+        let batches: Vec<Vec<Tup2<i32, Tup2<i32, i64>>>> = vec![
+            vec![Tup2(1, Tup2(2, 1))],
+            vec![Tup2(7, Tup2(8, 1))],
+            vec![Tup2(9, Tup2(10, 1))],
+            vec![Tup2(12, Tup2(12, 1))],
+            vec![Tup2(13, Tup2(13, 1))],
+        ];
+
+        generic_checkpoint_restore(batches, mkcircuit_with_bounds);
+    }
+
+    /// Make sure two circuits end up with a different fingerprint.
+    #[test]
+    fn fingerprint_is_different() {
+        let (_temp, cconf) = mkconfig();
+        let (mut dbsp, (_input_handle, _, _sample_size_handle)) = mkcircuit(&cconf).unwrap();
+        let fid1 = dbsp.fingerprint().unwrap();
+
+        let (_temp, cconf) = mkconfig();
+        let (mut dbsp, (_input_handle, _, _sample_size_handle)) =
+            mkcircuit_different(&cconf).unwrap();
+        let fid2 = dbsp.fingerprint().unwrap();
+        assert_ne!(fid1, fid2);
+
+        // Unfortunately, the fingerprint isn't perfect, e.g., it thinks these two
```

Comment:
we should be able to reduce this to a small value.
maybe the compiler should generate the fingerprint